--- a/externals/CMakeLists.txt	2022-04-08 13:36:05.937126207 -0400
+++ b/externals/CMakeLists.txt	2022-04-08 13:37:56.704182411 -0400
@@ -1,4 +1,5 @@
 # Definitions for all external bundled libraries
+include(FindPkgConfig)
 
 list(APPEND CMAKE_MODULE_PATH "${PROJECT_SOURCE_DIR}/CMakeModules")
 list(APPEND CMAKE_MODULE_PATH "${PROJECT_SOURCE_DIR}/externals/find-modules")
@@ -30,7 +31,11 @@
 add_subdirectory(glad)
 
 # inih
-add_subdirectory(inih)
+if (NOT USE_SYSTEM_INIH)
+    add_subdirectory(inih)
+else()
+    pkg_check_modules(INIH REQUIRED inih INIReader)
+endif()
 
 # mbedtls
 add_subdirectory(mbedtls EXCLUDE_FROM_ALL)
--- a/src/yuzu_cmd/config.cpp	2022-04-08 13:48:06.824622265 -0400
+++ b/src/yuzu_cmd/config.cpp	2022-04-08 13:49:46.411971812 -0400
@@ -16,7 +16,7 @@
 #pragma clang diagnostic pop
 #endif
 
-#include <inih/cpp/INIReader.h>
+#include <INIReader.h>
 #include "common/fs/file.h"
 #include "common/fs/fs.h"
 #include "common/fs/path_util.h"
--- a/src/yuzu_cmd/CMakeLists.txt       2022-04-06 13:20:33.149380529 -0400
+++ b/src/yuzu_cmd/CMakeLists.txt       2022-04-06 13:20:46.482348750 -0400
@@ -32,6 +32,9 @@
 if (MSVC)
     target_link_libraries(yuzu-cmd PRIVATE getopt)
 endif()
+if (USE_SYSTEM_INIH)
+    set(PLATFORM_LIBRARIES ${PLATFORM_LIBRARIES} ${INIH_LDFLAGS})
+endif()
 target_link_libraries(yuzu-cmd PRIVATE ${PLATFORM_LIBRARIES} SDL2 Threads::Threads)

 create_resource("../../dist/yuzu.bmp" "yuzu_cmd/yuzu_icon.h" "yuzu_icon")
