diff --git a/src/video_core/texture_cache/texture_cache.h b/src/video_core/texture_cache/texture_cache.h
index d575c57ca5e0..139f9df56c44 100644
--- a/src/video_core/texture_cache/texture_cache.h
+++ b/src/video_core/texture_cache/texture_cache.h
@@ -1574,9 +1574,6 @@ std::optional<typename TextureCache<P>::BlitImages> TextureCache<P>::GetBlitImag
                 if (!src_id && !dst_id) {
                     return std::nullopt;
                 }
-                if (src_id && True(slot_images[src_id].flags & ImageFlagBits::GpuModified)) {
-                    break;
-                }
                 if (dst_id && True(slot_images[dst_id].flags & ImageFlagBits::GpuModified)) {
                     break;
                 }
