--- src/music-players.h.org	2013-07-12 20:50:41.556784514 -0400
+++ src/music-players.h	2013-07-12 20:50:54.538783901 -0400
@@ -1836,6 +1836,8 @@
   // Reported by anonymous Sourceforge user
   { "Google Inc (for Barnes & Noble)", 0x18d1, "Nook Color", 0x2d02,
       DEVICE_FLAGS_ANDROID_BUGS },
+  { "Google Inc (for Barnes & Noble)", 0x18d1, "Nook Tablet", 0x6860,
+      DEVICE_FLAGS_ANDROID_BUGS },
   // Reported by anonymous Sourceforge user
   { "Google Inc (for Asus)", 0x18d1, "TF201 Transformer", 0x4d00,
       DEVICE_FLAGS_ANDROID_BUGS },
@@ -2189,6 +2189,13 @@
       DEVICE_FLAGS_ANDROID_BUGS },
 
   /*
+   * Kyocera
+   */
+  { "Kyocera", 0x0482, "Rise", 0x0571,
+      DEVICE_FLAGS_ANDROID_BUGS },
+  { "Kyocera", 0x0482, "Hydro", 0x0573,
+      DEVICE_FLAGS_ANDROID_BUGS },
+  /*
    * Hisense
    */
   // Reported by Anonymous SourceForge user
