diff --git a/src/core/hle/kernel/board/nintendo/nx/k_system_control.cpp b/src/core/hle/kernel/board/nintendo/nx/k_system_control.cpp
index 42d1fcc28269..2fc12de2a402 100644
--- a/src/core/hle/kernel/board/nintendo/nx/k_system_control.cpp
+++ b/src/core/hle/kernel/board/nintendo/nx/k_system_control.cpp
@@ -35,11 +35,11 @@ namespace {
 using namespace Common::Literals;
 
 u32 GetMemorySizeForInit() {
-    return Settings::values.use_extended_memory_layout ? Smc::MemorySize_6GB : Smc::MemorySize_4GB;
+    return Settings::values.use_extended_memory_layout ? Smc::MemorySize_8GB : Smc::MemorySize_4GB;
 }
 
 Smc::MemoryArrangement GetMemoryArrangeForInit() {
-    return Settings::values.use_extended_memory_layout ? Smc::MemoryArrangement_6GB
+    return Settings::values.use_extended_memory_layout ? Smc::MemoryArrangement_8GB
                                                        : Smc::MemoryArrangement_4GB;
 }
 } // namespace
diff --git a/src/yuzu/configuration/configure_general.ui b/src/yuzu/configuration/configure_general.ui
index 6cd79673c1ba..add110bb0203 100644
--- a/src/yuzu/configuration/configure_general.ui
+++ b/src/yuzu/configuration/configure_general.ui
@@ -64,7 +64,7 @@
           <item>
            <widget class="QCheckBox" name="use_extended_memory_layout">
             <property name="text">
-             <string>Extended memory layout (6GB DRAM)</string>
+             <string>Extended memory layout (8GB DRAM)</string>
             </property>
            </widget>
           </item>
diff --git a/src/yuzu_cmd/default_ini.h b/src/yuzu_cmd/default_ini.h
index 20e40340067b..209cfc28aec3 100644
--- a/src/yuzu_cmd/default_ini.h
+++ b/src/yuzu_cmd/default_ini.h
@@ -163,7 +163,7 @@ keyboard_enabled =
 # 0: Disabled, 1 (default): Enabled
 use_multi_core =
 
-# Enable extended guest system memory layout (6GB DRAM)
+# Enable extended guest system memory layout (8GB DRAM)
 # 0 (default): Disabled, 1: Enabled
 use_extended_memory_layout =
 
