--- a/externals/CMakeLists.txt
+++ b/externals/CMakeLists.txt
@@ -29,10 +30,7 @@
 add_subdirectory(catch2)
 
 # Crypto++
-set(CRYPTOPP_BUILD_DOCUMENTATION OFF)
-set(CRYPTOPP_BUILD_TESTING OFF)
-set(CRYPTOPP_INSTALL OFF)
-add_subdirectory(cryptopp-cmake)
+pkg_check_modules(CRYPTOPP REQUIRED libcryptopp)
 
 # HACK: The logic to set up the base include directory for CryptoPP does not work with Android SDK CMake 3.22.1.
 # Until there is a fixed version available, this code will detect and add in the proper include if it does not exist.
