--- libmpdemux/mpeg_hdr.c.org	2011-08-29 23:21:54.751742715 -0400
+++ libmpdemux/mpeg_hdr.c	2011-08-29 23:21:33.257742707 -0400
@@ -409,6 +409,10 @@
 {
   unsigned int n = 0, v, i, k, mbh;
   int frame_mbs_only;
+  unsigned char * _tmp_buf = malloc(len);
+
+  memcpy(_tmp_buf, buf, len);
+  buf = _tmp_buf;
 
   len = mp_unescape03(buf, len);
 
@@ -465,6 +469,9 @@
   if(getbits(buf, n++, 1))
     n = h264_parse_vui(picture, buf, n);
 
+  if (_tmp_buf)
+    free(_tmp_buf);
+
   return n;
 }
 
