--- cfg-common.h.org	2010-10-06 22:45:34.854836882 -0400
+++ cfg-common.h	2010-10-06 22:46:11.520440128 -0400
@@ -559,6 +559,8 @@
     {"xvidopts", xvid_dec_opts, CONF_TYPE_SUBCONFIG, 0, 0, 0, NULL},
 #endif
     {"codecs-file", &codecs_file, CONF_TYPE_STRING, 0, 0, 0, NULL},
+    //Auto-detect AC3 channels with -af lavcac3enc
+    {"lavc_ac3_detect", &lavc_ac3_detect, CONF_TYPE_INT, 0, 0, 0, NULL},
 // ------------------------- subtitles options --------------------
 
     {"sub", &sub_name, CONF_TYPE_STRING_LIST, 0, 0, 0, NULL},
--- libmpcodecs/dec_audio.h.org	2010-10-06 22:47:17.395913946 -0400
+++ libmpcodecs/dec_audio.h	2010-10-06 22:47:52.610376612 -0400
@@ -26,6 +26,8 @@
 
 extern int audio_output_channels;
 extern int fakemono;
+extern int lavc_ac3_detect;
+extern int lavc_ac3_num_detected;
 
 // dec_audio.c:
 void afm_help(void);
--- mplayer.c.orig	2010-10-06 22:41:29.972771500 -0400
+++ mplayer.c	2010-10-06 22:50:19.792838682 -0400
@@ -140,6 +140,10 @@
 #endif
 
 
+extern int lavc_ac3_detect;
+extern int lavc_ac3_num_detected;
+extern int audio_output_channels;
+
 int slave_mode=0;
 int player_idle_mode=0;
 int quiet=0;
@@ -573,6 +577,11 @@
     mp_msg(MSGT_IDENTIFY,MSGL_INFO,"ID_AUDIO_BITRATE=%d\n", mpctx->sh_audio->i_bps*8);
     mp_msg(MSGT_IDENTIFY,MSGL_INFO,"ID_AUDIO_RATE=%d\n", mpctx->sh_audio->samplerate);
     mp_msg(MSGT_IDENTIFY,MSGL_INFO,"ID_AUDIO_NCH=%d\n", mpctx->sh_audio->channels);
+    if (lavc_ac3_detect) {
+      mp_msg(MSGT_IDENTIFY,MSGL_INFO,"Autodetecting number of channels=%d\n", mpctx->sh_audio->channels);
+      audio_output_channels = mpctx->sh_audio->channels;
+      lavc_ac3_num_detected = mpctx->sh_audio->channels;
+    }
     start_pts = ds_get_next_pts(mpctx->d_audio);
   }
   if (video_start_pts != MP_NOPTS_VALUE) {
--- libmpcodecs/dec_audio.c.orig	2010-01-30 11:57:40.000000000 -0500
+++ libmpcodecs/dec_audio.c	2010-05-07 15:23:40.096774562 -0400
@@ -48,6 +48,8 @@
 /* used for ac3surround decoder - set using -channels option */
 int audio_output_channels = 2;
 af_cfg_t af_cfg = { 1, NULL };	// Configuration for audio filters
+extern int lavc_ac3_detect;
+extern int lavc_ac3_num_detected;
 
 void afm_help(void)
 {
@@ -132,6 +134,10 @@
     mp_msg(MSGT_IDENTIFY, MSGL_INFO,
 	   "ID_AUDIO_BITRATE=%d\nID_AUDIO_RATE=%d\n" "ID_AUDIO_NCH=%d\n",
 	   sh_audio->i_bps * 8, sh_audio->samplerate, sh_audio->channels);
+    if ((lavc_ac3_detect) && (lavc_ac3_num_detected == 0)) {
+        lavc_ac3_num_detected = sh_audio->channels;
+        audio_output_channels = sh_audio->channels;
+    }
 
     sh_audio->a_out_buffer_size = 0;
     sh_audio->a_out_buffer = NULL;
--- libaf/af_lavcac3enc.c.orig	2010-10-06 22:50:05.667450541 -0400
+++ libaf/af_lavcac3enc.c	2010-10-06 22:50:19.794838878 -0400
@@ -35,6 +35,9 @@
 #include "libavcodec/ac3.h"
 #include "libavutil/intreadwrite.h"
 
+int lavc_ac3_detect = 0;
+int lavc_ac3_num_detected = 0;
+
 // Data for specific instances of this filter
 typedef struct af_ac3enc_s {
     struct AVCodec        *lavc_acodec;
@@ -130,6 +130,12 @@
         }
         if (s->min_channel_num == 0)
             s->min_channel_num = 5;
+
+        if (lavc_ac3_detect) {
+            mp_msg(MSGT_AFILTER, MSGL_V, "af_lavcac3enc using auto-detect on channels: %d\n", lavc_ac3_num_detected);
+            s->min_channel_num = lavc_ac3_num_detected;
+        }
+
         mp_msg(MSGT_AFILTER, MSGL_V, "af_lavcac3enc config spdif:%d, bitrate:%d, "
                "minchnum:%d.\n", s->add_iec61937_header, s->bit_rate,
                s->min_channel_num);
