diff --git a/.ci/scripts/linux/docker.sh b/.ci/scripts/linux/docker.sh
index c8bc56c9a597..7f6d2ad1be8c 100755
--- a/.ci/scripts/linux/docker.sh
+++ b/.ci/scripts/linux/docker.sh
@@ -22,6 +22,7 @@ cmake .. \
       -DUSE_DISCORD_PRESENCE=ON \
       -DYUZU_ENABLE_COMPATIBILITY_REPORTING=${ENABLE_COMPATIBILITY_REPORTING:-"OFF"} \
       -DYUZU_USE_BUNDLED_FFMPEG=ON \
+      -DYUZU_ENABLE_LTO=ON \
       -GNinja
 
 ninja
