From aaddaf5edd0c1b99e4d6761e0283fab5c7a8e0f5 Mon Sep 17 00:00:00 2001
From: James Harris <James.R.Harris@gmail.com>
Date: Tue, 20 Jun 2017 19:03:00 -0700
Subject: [PATCH] Support kernel 4.11.0

---
 ctn91xx_kal.h | 4 ++++
 ctn91xx_net.c | 4 ++++
 2 files changed, 8 insertions(+)

diff --git a/ctn91xx_kal.h b/ctn91xx_kal.h
index b6c8fb6..ce633f3 100644
--- a/ctn91xx_kal.h
+++ b/ctn91xx_kal.h
@@ -27,6 +27,10 @@
 #include <linux/cdev.h>
 #include <linux/list.h>
 
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(4,11,0)
+#include <linux/sched/signal.h>
+#endif
+
 //it's in an arch specific header...this is just easier
 #define TCGETS 0x5401
 
diff --git a/ctn91xx_net.c b/ctn91xx_net.c
index 5b9bbf0..de0702e 100644
--- a/ctn91xx_net.c
+++ b/ctn91xx_net.c
@@ -242,7 +242,9 @@ void ctn91xx_net_rx_skb( ctn91xx_dev_t* dev, struct sk_buff* skb, uint16_t rx_le
     skb->protocol = eth_type_trans( skb, netdev );
     netif_rx( skb );
 
+#if LINUX_VERSION_CODE < KERNEL_VERSION(4,11,0)
     netdev->last_rx = jiffies;
+#endif
     priv->stats.rx_bytes += rx_len;
     priv->stats.rx_packets++;
 
@@ -287,7 +289,9 @@ irqreturn_t ctn91xx_net_isr(int irq, void *ptr)
                 skb->protocol = eth_type_trans( skb, netdev );
                 netif_rx( skb );
 
+#if LINUX_VERSION_CODE < KERNEL_VERSION(4,11,0)
                 netdev->last_rx = jiffies;
+#endif
                 priv->stats.rx_bytes += rx_len;
                 priv->stats.rx_packets++;
             } else {
