diff --git a/ctn91xx_driver.c b/ctn91xx_driver.c
index 062d1f3..54bcdf2 100644
--- a/ctn91xx_driver.c
+++ b/ctn91xx_driver.c
@@ -129,7 +129,7 @@ static int ctn91xx_mmap(struct file* filp, struct vm_area_struct* vma)
 #if LINUX_VERSION_CODE < KERNEL_VERSION(3,7,0)
             vma->vm_flags |= VM_IO | VM_RESERVED;
 #else
-            vma->vm_flags |= VM_IO | VM_DONTEXPAND | VM_DONTDUMP;
+            vm_flags_set(vma, vma->vm_flags | VM_IO | VM_DONTEXPAND | VM_DONTDUMP);
 #endif
 
             if(io_remap_pfn_range(vma, vma->vm_start,
@@ -151,7 +151,7 @@ static int ctn91xx_mmap(struct file* filp, struct vm_area_struct* vma)
 #if LINUX_VERSION_CODE < KERNEL_VERSION(3,7,0)
             vma->vm_flags |= VM_IO | VM_RESERVED;
 #else
-            vma->vm_flags |= VM_IO | VM_DONTEXPAND | VM_DONTDUMP;
+            vm_flags_set(vma, vma->vm_flags | VM_IO | VM_DONTEXPAND | VM_DONTDUMP);
 #endif
 
             if(io_remap_pfn_range(vma, vma->vm_start,
