Index: vdr-1.7.7/videodir.c
===================================================================
--- vdr-1.7.7.orig/videodir.c
+++ vdr-1.7.7/videodir.c
@@ -196,9 +196,10 @@ bool RemoveVideoFile(const char *FileNam
 #ifdef USE_HARDLINKCUTTER
 static bool StatNearestDir(const char *FileName, struct stat *Stat)
 {
-  cString Name(FileName);
+  char *Name = strdup(FileName);
+  cString Name_str(Name, true); /* manage free for us */
   char *p;
-  while ((p = strrchr((const char*)Name + 1, '/')) != NULL) {
+  while ((p = strrchr(Name + 1, '/')) != NULL) {
         *p = 0; // truncate at last '/'
         if (stat(Name, Stat) == 0) {
            isyslog("StatNearestDir: Stating %s", (const char*)Name);
