diff -Naur vdr-1.7.4.orig/dvbdevice.c vdr-1.7.4/dvbdevice.c
--- vdr-1.7.4.orig/dvbdevice.c	2009-02-03 21:21:34.000000000 +0100
+++ vdr-1.7.4/dvbdevice.c	2009-02-03 21:26:11.000000000 +0100
@@ -333,14 +333,14 @@
      }
 #ifdef USE_ATSC
   else if (frontendType == SYS_ATSC) {
-     // DVB-T
+     // ATSC
      SETCMD(DTV_DELIVERY_SYSTEM, frontendType);
      SETCMD(DTV_FREQUENCY, FrequencyToHz(channel.Frequency()));
      SETCMD(DTV_INVERSION, channel.Inversion());
      SETCMD(DTV_MODULATION, channel.Modulation());
 
-     tuneTimeout = DVBC_TUNE_TIMEOUT;
-     lockTimeout = DVBC_LOCK_TIMEOUT;
+     tuneTimeout = DVBT_TUNE_TIMEOUT;
+     lockTimeout = DVBT_LOCK_TIMEOUT;
      }
 #endif /* ATSC */
   else if (frontendType == SYS_DVBT) {
@@ -922,12 +922,12 @@
 #endif /* SOURCECAPS */
   return type == cSource::stNone
       || type == cSource::stCable && (frontendType == SYS_DVBC_ANNEX_AC || frontendType == SYS_DVBC_ANNEX_B)
-#ifdef USE_ATSC
-      || type == cSource::stCable && (frontendType == SYS_ATSC)
-      || type == cSource::stTerr  && (frontendType == SYS_ATSC)
-#endif /* ATSC */
       || type == cSource::stSat   && (frontendType == SYS_DVBS || frontendType == SYS_DVBS2)
+#ifdef USE_ATSC
+      || type == cSource::stTerr  && (frontendType == SYS_DVBT || frontendType == SYS_ATSC);
+#else
       || type == cSource::stTerr  && (frontendType == SYS_DVBT);
+#endif /* ATSC */
 }
 
 bool cDvbDevice::ProvidesTransponder(const cChannel *Channel) const
