--- dvbdevice.c.org	2009-05-20 11:16:51.000000000 -0400
+++ dvbdevice.c	2009-05-20 11:18:05.000000000 -0400
@@ -748,18 +748,23 @@
 {
   eVideoSystem VideoSystem = vsPAL;
   video_size_t vs;
+  if (fd_video >= 0)
+  {
   if (ioctl(fd_video, VIDEO_GET_SIZE, &vs) == 0) {
      if (vs.h == 480 || vs.h == 240)
         VideoSystem = vsNTSC;
      }
   else
      LOG_ERROR;
+  }
   return VideoSystem;
 }
 
 void cDvbDevice::GetVideoSize(int &Width, int &Height, eVideoAspect &Aspect)
 {
   video_size_t vs;
+  if (fd_video >= 0)
+  {
   if (ioctl(fd_video, VIDEO_GET_SIZE, &vs) == 0) {
      Width = vs.w;
      if (Width < 720) // FIXME: some channels result in a With of, e.g. 544, but the final video *is* 720 wide
@@ -771,6 +776,7 @@
      }
   else
      LOG_ERROR;
+  }
   cDevice::GetVideoSize(Width, Height, Aspect);
 }
 
