diff --git a/src/core/hle/kernel/physical_core.cpp b/src/core/hle/kernel/physical_core.cpp
index 7fa8e2a85248..0f45a324946f 100644
--- a/src/core/hle/kernel/physical_core.cpp
+++ b/src/core/hle/kernel/physical_core.cpp
@@ -139,7 +139,7 @@ void PhysicalCore::RunThread(Kernel::KThread* thread) {
         }
 
         // Handle external interrupt sources.
-        if (interrupt || !m_is_single_core) {
+        if (interrupt || m_is_single_core) {
             return;
         }
     }
