--- src/main.c	2008-10-25 02:02:09.000000000 -0400
+++ src.new/main.c	2008-10-25 02:04:46.000000000 -0400
@@ -137,8 +137,10 @@
 int main (int argc, char *argv[])
 {
    int result;
+   int argument;
    BOOL want_exit = FALSE;
    BOOL enter_gui = TRUE;
+   BOOL no_gui = FALSE;
 
    /* Clear the console. */
    console_clear ();
@@ -184,7 +186,15 @@
 
    if (argc >= 2)
    {
-      if (load_rom (argv[1], &global_rom) != 0)
+      for (argument = 0; argument < argc; argument++)
+      {
+        if (!strcmp(argv[argument], "-nogui"))
+	{
+	  printf("No Gui!\n");
+	  no_gui = TRUE;
+	}
+      }
+      if (load_rom (argv[argc-1], &global_rom) != 0)
       {
          WARN("Failed to load ROM (bad format?)");
 
@@ -264,7 +274,8 @@
    {
       if (enter_gui)
       {
-         want_exit = show_gui (first_run);
+	if (no_gui) { want_exit = TRUE; }
+	else { want_exit = show_gui (first_run); }
 
          if (first_run)
             first_run = FALSE;
