*** old/MoBlock-0.8/Makefile	2006-03-22 18:44:31.000000000 +0100
--- new/MoBlock-0.8/Makefile	2009-04-08 11:21:11.000000000 +0200
***************
*** 7,15 ****
  #QUEUE_LIB=LIBIPQ
  QUEUE_LIB=NFQUEUE
  
! CFLAGS=-Wall -O2 -march=i586 -mtune=i686 -fomit-frame-pointer -ffast-math \
! 	-D_GNU_SOURCE -D$(QUEUE_LIB) -L/usr/include/libipq
! CC=gcc
  
  all: moblock
  
--- 7,13 ----
  #QUEUE_LIB=LIBIPQ
  QUEUE_LIB=NFQUEUE
  
! CFLAGS+=-Wall -D_GNU_SOURCE -D$(QUEUE_LIB) -L/usr/include/libipq
  
  all: moblock
  
***************
*** 17,23 ****
  moblock: MoBlock.o rbt.o
  	gcc -o $@ MoBlock.o rbt.o -lnetfilter_queue -lnfnetlink
  	#gcc -o $@ MoBlock.o rbt.o -lipq
- 	strip $@
  
  moblock-static: MoBlock.o rbt.o
  	gcc -static -o $@ MoBlock.o rbt.o -lnetfilter_queue -lnfnetlink
--- 15,20 ----
***************
*** 28,33 ****
  	rm -f *.o *~ *# moblock
  
  install:
! 	install -m 755 moblock $(DESTDIR)/usr/bin
  
  .PHONY: clean
--- 25,31 ----
  	rm -f *.o *~ *# moblock
  
  install:
! 	mkdir -p "$(DESTDIR)/usr/bin"
! 	install -m 755 moblock "$(DESTDIR)/usr/bin"
  
  .PHONY: clean
*** old/MoBlock-0.8/MoBlock.c	2006-03-22 18:44:31.000000000 +0100
--- new/MoBlock-0.8/MoBlock.c	2009-04-08 11:15:56.000000000 +0200
***************
*** 32,38 ****
--- 32,40 ----
  #include <netinet/udp.h>
  #include <sys/socket.h>
  #include <arpa/inet.h>
+ #include <limits.h>
  #include <linux/netfilter_ipv4.h>
+ #include <linux/version.h>
  #include <signal.h>
  #include <regex.h>
  
***************
*** 313,319 ****
  			reopen_logfile();
  			break;
  		case SIGUSR2:
! 			fprintf(logfile,"Got SIGUSR2! Dumping stats to /var/log/MoBlock.stats\n");
  			ll_log();
  			break;
  		case SIGHUP:
--- 315,321 ----
  			reopen_logfile();
  			break;
  		case SIGUSR2:
! 			fprintf(logfile,"Got SIGUSR2! Dumping stats to /var/log/moblock.stats\n");
  			ll_log();
  			break;
  		case SIGHUP:
***************
*** 502,508 ****
--- 504,512 ----
  
  	if (nfq_unbind_pf(h, AF_INET) < 0) {
  		fprintf(logfile, "error during nfq_unbind_pf()\n");
+ #if LINUX_VERSION_CODE < KERNEL_VERSION(2, 6, 23)
  		exit(-1);
+ #endif
  	}
  
  	if (nfq_bind_pf(h, AF_INET) < 0) {
*** old/MoBlock-0.8/rbt.c	2006-03-22 18:44:31.000000000 +0100
--- new/MoBlock-0.8/rbt.c	2009-04-08 11:03:55.000000000 +0200
***************
*** 144,152 ****
      FILE *fp;
      time_t tp;
      
!     fp=fopen("/var/log/MoBlock.stats","a");
      if ( fp == NULL ) {
!         fprintf(stderr,"Error opening stats file /var/log/MoBlock.stats\n");
          perror("ll_log");
          return;
      }
--- 144,152 ----
      FILE *fp;
      time_t tp;
      
!     fp=fopen("/var/log/moblock.stats","a");
      if ( fp == NULL ) {
!         fprintf(stderr,"Error opening stats file /var/log/moblock.stats\n");
          perror("ll_log");
          return;
      }
***************
*** 160,166 ****
      }
      fprintf(fp,"----------------------------------------\n");
      if ( fclose(fp) != 0 ) {
!         perror("Error closing stats file /var/log/MoBlock.stats");
          return;
      }
  }
--- 160,166 ----
      }
      fprintf(fp,"----------------------------------------\n");
      if ( fclose(fp) != 0 ) {
!         perror("Error closing stats file /var/log/moblock.stats");
          return;
      }
  }
*** old/MoBlock-0.8/README	2006-03-22 18:44:31.000000000 +0100
--- new/MoBlock-0.8/README	2009-04-08 11:03:55.000000000 +0200
***************
*** 145,155 ****
  
     kill -TERM <MoBlockPid>
  
!    While shutting down it will dump some stats to /var/log/MoBlock.stats
     To obtain stats about blocked ranges while it's running:
  
     kill -USR1 <MoBlockPid>	# write stats to logfile
!    kill -USR2 <MoBlockPid>      # write stats to /var/log/MoBlock.stats
  
  ** NEW: to reload the blocklist while MoBlock is running send to it the
     HUP signal:
--- 145,155 ----
  
     kill -TERM <MoBlockPid>
  
!    While shutting down it will dump some stats to /var/log/moblock.stats
     To obtain stats about blocked ranges while it's running:
  
     kill -USR1 <MoBlockPid>	# write stats to logfile
!    kill -USR2 <MoBlockPid>      # write stats to /var/log/moblock.stats
  
  ** NEW: to reload the blocklist while MoBlock is running send to it the
     HUP signal:
