diff -Nur DarwinStreamingSrvr6.0.3-Source/Buildit DarwinStreamingSrvr6.0.3-Source.x86_64-m32/Buildit
--- DarwinStreamingSrvr6.0.3-Source/Buildit	2008-06-05 12:29:29.000000000 -0500
+++ DarwinStreamingSrvr6.0.3-Source.x86_64-m32/Buildit	2008-06-05 12:30:22.000000000 -0500
@@ -116,7 +116,7 @@
 		LINKER='gcc'
  		MAKE=make
  		
-		COMPILER_FLAGS="-D_REENTRANT -D__USE_POSIX -D__linux__ -pipe -fPIC"
+		COMPILER_FLAGS="-D_REENTRANT -D__USE_POSIX -D__linux__ -pipe -fPIC -m32"
 		INCLUDE_FLAG="-include"
 		
 		CORE_LINK_LIBS="-lpthread -ldl -lstdc++ -lm -lcrypt"
