diff -Naur DarwinStreamingSrvr6.0.3-Source/APIStubLib/BuildAPIStubLib DarwinStreamingSrvr6.0.3-Source.patched/APIStubLib/BuildAPIStubLib
--- DarwinStreamingSrvr6.0.3-Source/APIStubLib/BuildAPIStubLib	2002-09-09 19:58:14.000000000 -0500
+++ DarwinStreamingSrvr6.0.3-Source.patched/APIStubLib/BuildAPIStubLib	2007-07-16 09:02:31.000000000 -0500
@@ -2,6 +2,9 @@
 
 x=`uname`
 y=`uname -m`
+if [ "$y" == "ppc64" ]; then
+	y="i686"
+fi
 
 PLAT=$x.$y
 
diff -Naur DarwinStreamingSrvr6.0.3-Source/Buildit DarwinStreamingSrvr6.0.3-Source.patched/Buildit
--- DarwinStreamingSrvr6.0.3-Source/Buildit	2006-07-14 04:36:35.000000000 -0500
+++ DarwinStreamingSrvr6.0.3-Source.patched/Buildit	2007-07-16 09:03:07.000000000 -0500
@@ -55,6 +55,9 @@
 
 OSNAME=`uname`
 HARDWARENAME=`uname -m`
+if [ "$HARDWARENAME" == "ppc64" ]; then
+	HARDWARENAME="i686"
+fi
  
 PLAT=$OSNAME.$HARDWARENAME
 
diff -Naur DarwinStreamingSrvr6.0.3-Source/PlaylistBroadcaster.tproj/BuildPlaylist DarwinStreamingSrvr6.0.3-Source.patched/PlaylistBroadcaster.tproj/BuildPlaylist
--- DarwinStreamingSrvr6.0.3-Source/PlaylistBroadcaster.tproj/BuildPlaylist	2002-09-09 20:07:18.000000000 -0500
+++ DarwinStreamingSrvr6.0.3-Source.patched/PlaylistBroadcaster.tproj/BuildPlaylist	2007-07-16 09:03:28.000000000 -0500
@@ -2,6 +2,9 @@
 
 x=`uname`
 y=`uname -m`
+if [ "$y" == "ppc64" ]; then
+	y="i686"
+fi
 
 PLAT=$x.$y
 
diff -Naur DarwinStreamingSrvr6.0.3-Source/StreamingProxy.tproj/BuildProxy DarwinStreamingSrvr6.0.3-Source.patched/StreamingProxy.tproj/BuildProxy
--- DarwinStreamingSrvr6.0.3-Source/StreamingProxy.tproj/BuildProxy	2001-11-13 16:46:00.000000000 -0600
+++ DarwinStreamingSrvr6.0.3-Source.patched/StreamingProxy.tproj/BuildProxy	2007-07-16 09:03:42.000000000 -0500
@@ -2,6 +2,9 @@
 
 x=`uname`
 y=`uname -m`
+if [ "$y" == "ppc64" ]; then
+	y="i686"
+fi
 
 PLAT=$x.$y
 
