--- stream/stream_ffmpeg.c.org	2012-11-05 00:02:55.000000000 -0500
+++ stream/stream_ffmpeg.c	2012-11-05 00:03:48.000000000 -0500
@@ -124,6 +124,14 @@
     if (!dummy && avio_open(&avio, filename, flags) < 0)
         goto out;
 
+    if (!strncmp(filename, "applehttp:", 10)) {
+        char *newname = strdup(&stream->url[5]);
+        if (newname) {
+            free(stream->url);
+            filename = stream->url = newname;
+        }
+    }
+
     char *rtmp[] = {"rtmp:", "rtmpt:", "rtmpe:", "rtmpte:", "rtmps:"};
     for (int i = 0; i < FF_ARRAY_ELEMS(rtmp); i++)
         if (!strncmp(filename, rtmp[i], strlen(rtmp[i]))) {
@@ -158,7 +166,7 @@
   "",
   "",
   open_f,
-  { "ffmpeg", "rtmp", NULL },
+  { "ffmpeg", "rtmp", "applehttp", NULL },
   NULL,
   1 // Urls are an option string
 };
--- libmpdemux/demux_lavf.c.org	2012-11-05 00:03:02.000000000 -0500
+++ libmpdemux/demux_lavf.c	2012-11-05 00:04:41.000000000 -0500
@@ -615,8 +615,11 @@
         if (!strncmp(demuxer->stream->url, "ffmpeg://rtsp:", 14))
             av_strlcpy(mp_filename, demuxer->stream->url + 9,
                        sizeof(mp_filename));
-        else
+        else {
+            if (!strncmp("http:", demuxer->stream->url, 5))
+               memset(mp_filename, 0, 256);
             av_strlcat(mp_filename, demuxer->stream->url, sizeof(mp_filename));
+        }
     } else
         av_strlcat(mp_filename, "foobar.dummy", sizeof(mp_filename));
 
