Index: libavcodec/avcodec.h
===================================================================
--- libavcodec/avcodec.h	(revision 19235)
+++ libavcodec/avcodec.h	(working copy)
@@ -400,7 +400,7 @@
 #define CH_LAYOUT_STEREO_DOWNMIX    (CH_STEREO_LEFT|CH_STEREO_RIGHT)
 
 /* in bytes */
-#define AVCODEC_MAX_AUDIO_FRAME_SIZE 192000 // 1 second of 48khz 32bit audio
+#define AVCODEC_MAX_AUDIO_FRAME_SIZE (2<<20)
 
 /**
  * Required number of additionally allocated bytes at the end of the input bitstream for decoding.
