diff --git a/src/hid_core/frontend/emulated_controller.cpp b/src/hid_core/frontend/emulated_controller.cpp
index 2ab93402d71e..b001cf2fdeb0 100644
--- a/src/hid_core/frontend/emulated_controller.cpp
+++ b/src/hid_core/frontend/emulated_controller.cpp
@@ -110,6 +110,7 @@ void EmulatedController::ReloadFromSettings() {
         original_npad_type = npad_type;
     }
 
+    SetPollingMode(EmulatedDeviceIndex::RightIndex, Common::Input::PollingMode::Active);
     Disconnect();
     if (player.connected) {
         Connect();
