--- cfg-mplayer.h.org	2010-10-29 22:29:08.693000090 -0400
+++ cfg-mplayer.h	2010-10-29 22:30:03.417000090 -0400
@@ -38,6 +38,8 @@
 #include "unrar_exec.h"
 
 
+extern int growing_file_pause_enable;
+
 const m_option_t vd_conf[]={
     {"help", "Use MPlayer with an appropriate video file instead of live partners to avoid vd.\n", CONF_TYPE_PRINT, CONF_NOCFG|CONF_GLOBAL, 0, 0, NULL},
     {NULL, NULL, 0, 0, 0, 0, NULL}
@@ -346,6 +348,8 @@
     {"tvscan", "MPlayer was compiled without TV interface support.\n", CONF_TYPE_PRINT, 0, 0, 0, NULL},
 #endif /* CONFIG_TV */
 
+    {"livepause", &growing_file_pause_enable, CONF_TYPE_INT, CONF_RANGE, 0, 60, NULL},
+
     {"list-properties", &list_properties, CONF_TYPE_FLAG, CONF_GLOBAL, 0, 1, NULL},
     {"identify", &mp_msg_levels[MSGT_IDENTIFY], CONF_TYPE_FLAG, CONF_GLOBAL, 0, MSGL_V, NULL},
     {"-help", help_text, CONF_TYPE_PRINT, CONF_NOCFG|CONF_GLOBAL, 0, 0, NULL},
--- libmpdemux/demuxer.c.org	2010-10-29 22:29:14.463000091 -0400
+++ libmpdemux/demuxer.c	2010-10-29 22:42:09.995000097 -0400
@@ -35,6 +35,7 @@
 #include "libvo/fastmemcpy.h"
 
 #include "stream/stream.h"
+#include "stream/cache2.h"
 #include "demuxer.h"
 #include "stheader.h"
 #include "mf.h"
@@ -190,6 +191,10 @@
     NULL
 };
 
+
+double first_pts_hack = 0, last_pts_update = 0;
+int growing_file_pause_enable=0;
+
 void free_demuxer_stream(demux_stream_t *ds)
 {
     ds_free_packs(ds);
@@ -646,6 +651,22 @@
             if (p->pts != MP_NOPTS_VALUE) {
                 ds->pts = p->pts;
                 ds->pts_bytes = 0;
+
+                if (!last_pts_update)
+                    last_pts_update = p->pts;
+
+                if (growing_file_pause_enable && (p->pts - last_pts_update > 3))
+                {
+                  last_pts_update = p->pts;
+                  if (demux->stream->cache_data)
+                  {
+                    off_t *m = cache_get_lpause(demux->stream->cache_data);
+                    if (m)
+                      demux->stream->end_pos = *m;
+                  }
+                  if (demux->stream->end_pos > demux->movi_end)
+                    demux->movi_end = demux->stream->end_pos;
+                }
             }
             ds->pts_bytes += p->len;    // !!!
             if (p->stream_pts != MP_NOPTS_VALUE)
@@ -883,6 +904,23 @@
         if (!demux_fill_buffer(demux, ds))
             return MP_NOPTS_VALUE;
     }
+    if (growing_file_pause_enable)
+    {
+      if (ds && ds->first)
+      {
+           if (first_pts_hack == 0)
+               first_pts_hack = ds->first->pts;
+           else if (ds->first->pts <  first_pts_hack)
+               first_pts_hack = ds->first->pts;
+      }
+      if (ds && ds->current)
+      {
+           if (first_pts_hack == 0)
+               first_pts_hack = ds->current->pts;
+           else if (ds->first->pts <  first_pts_hack)
+               first_pts_hack = ds->current->pts;
+      }
+    }
     // take pts from "current" if we never read from it.
     if (ds->current && !ds->buffer_pos)
         return ds->current->pts;
@@ -1450,6 +1488,11 @@
         get_time_ans = demuxer->stream_pts;
     else if (sh_video)
         get_time_ans = sh_video->pts;
+    if (growing_file_pause_enable)
+    {
+      if (get_time_ans - first_pts_hack > 0)
+          return (int)get_time_ans - first_pts_hack;
+    }
     return (int) get_time_ans;
 }
 
--- mplayer.c.org	2010-10-29 22:29:02.528000091 -0400
+++ mplayer.c	2010-10-29 22:30:03.416000090 -0400
@@ -148,6 +148,9 @@
 double start_pts = MP_NOPTS_VALUE;
 char *heartbeat_cmd;
 
+extern double first_pts_hack;
+extern int growing_file_pause_enable;
+
 #define ROUND(x) ((int)((x)<0 ? (x)-0.5 : (x)+0.5))
 
 
@@ -2677,6 +2680,33 @@
 static int seek(MPContext *mpctx, double amount, int style)
 {
     current_module = "seek";
+
+    if (growing_file_pause_enable)
+    {
+      if (amount > 0 && mpctx->demuxer->movi_end > mpctx->demuxer->filepos)
+      {
+        sh_video_t *sh_video = mpctx->demuxer->video->sh; amount = 7;
+        if (sh_video)
+        {
+            double new_pts = sh_video->pts - first_pts_hack;
+            double seek_1_second = ((double)mpctx->demuxer->filepos / new_pts);
+            double size_now_to_end = (double)(mpctx->demuxer->movi_end - mpctx->demuxer->filepos);
+            double time_to_end = size_now_to_end/seek_1_second;
+
+//            printf("..... time_to_end: %0.2f -------- \n", time_to_end);
+            if (time_to_end < (3 + growing_file_pause_enable))
+            {
+                set_osd_msg(OSD_MSG_TEXT, 1, osd_duration, "LIVE: %0.2f", time_to_end);
+                return -1;
+            }
+            if (time_to_end < amount + 3 + growing_file_pause_enable)
+                amount = time_to_end - (5 + growing_file_pause_enable);
+//            printf("..... amount to ff: %0.2f -------- \n", amount);
+
+        }
+      }
+    }
+
     if (demux_seek(mpctx->demuxer, amount, audio_delay, style) == 0)
 	return -1;
 
--- stream/cache2.c.org	2010-10-29 22:29:27.521000090 -0400
+++ stream/cache2.c	2010-10-29 22:30:03.437000090 -0400
@@ -80,6 +80,7 @@
   off_t offset;      // filepos <-> bufferpos  offset value (filepos of the buffer's first byte)
   // reader's pointers:
   off_t read_filepos;
+  off_t lpause_off_hack;
   // commands/locking:
 //  int seek_lock;   // 1 if we will seek/reset buffer, 2 if we are ready for cmd
 //  int fifo_flag;  // 1 if we should use FIFO to notice cache about buffer reads.
@@ -97,6 +98,17 @@
 
 int cache_fill_status=0;
 
+off_t * cache_get_lpause(void * s)
+{
+  cache_vars_t *c = (cache_vars_t*)s;
+  off_t * ret = NULL;
+  if (s && c)
+  {
+     ret = &c->lpause_off_hack;
+  }
+  return ret;
+}
+
 static void cache_wakeup(stream_t *s)
 {
 #if FORKED_CACHE
--- stream/cache2.h.org	2010-10-29 22:29:34.480000091 -0400
+++ stream/cache2.h	2010-10-29 22:30:03.438000090 -0400
@@ -25,5 +25,6 @@
 
 void cache_uninit(stream_t *s);
 int cache_do_control(stream_t *stream, int cmd, void *arg);
+off_t *  cache_get_lpause(void *);
 
 #endif /* MPLAYER_CACHE2_H */
--- stream/stream_file.c.org	2010-10-29 22:29:20.125000092 -0400
+++ stream/stream_file.c	2010-10-29 22:30:03.429000090 -0400
@@ -26,6 +26,7 @@
 
 #include "mp_msg.h"
 #include "stream.h"
+#include "cache2.h"
 #include "help_mp.h"
 #include "m_option.h"
 #include "m_struct.h"
@@ -51,8 +52,25 @@
   stream_opts_fields
 };
 
+static int control(stream_t *s, int cmd, void *arg);
+extern int growing_file_pause_enable;
+
 static int fill_buffer(stream_t *s, char* buffer, int max_len){
   int r = read(s->fd,buffer,max_len);
+  if (growing_file_pause_enable)
+  {
+    off_t sz = 0;
+    control(s, STREAM_CTRL_GET_SIZE, &sz);
+    if (sz > s->end_pos)
+        s->end_pos = sz;
+    if (s->cache_data)
+    {
+       //void * moo = s->cache_data;
+       off_t * moo = cache_get_lpause(s->cache_data);
+       if (moo)
+           memcpy(moo, &sz,sizeof(off_t)); 
+    }
+  }
   return (r <= 0) ? -1 : r;
 }
 
