--- libavformat/utils.c.org	2010-12-26 16:32:54.000000000 -0500
+++ libavformat/utils.c	2010-12-26 16:33:07.000000000 -0500
@@ -2836,7 +2836,7 @@
         av_log(s, AV_LOG_ERROR,
                "st:%d error, non monotone timestamps %"PRId64" >= %"PRId64"\n",
                st->index, st->cur_dts, pkt->dts);
-        return -1;
+        //return -1;
     }
     if(pkt->dts != AV_NOPTS_VALUE && pkt->pts != AV_NOPTS_VALUE && pkt->pts < pkt->dts){
         av_log(s, AV_LOG_ERROR, "st:%d error, pts < dts\n", st->index);
