Index: libavcodec/allcodecs.c
===================================================================
--- libavcodec/allcodecs.c	(revision 19340)
+++ libavcodec/allcodecs.c	(working copy)
@@ -233,6 +233,7 @@
     REGISTER_DECODER (WAVPACK, wavpack);
     REGISTER_ENCDEC  (WMAV1, wmav1);
     REGISTER_ENCDEC  (WMAV2, wmav2);
+    REGISTER_DECODER (WMAPRO, wmapro);
     REGISTER_DECODER (WS_SND1, ws_snd1);
 
     /* PCM codecs */
Index: libavcodec/Makefile
===================================================================
--- libavcodec/Makefile	(revision 19340)
+++ libavcodec/Makefile	(working copy)
@@ -259,6 +259,7 @@
 OBJS-$(CONFIG_WMAV1_ENCODER)           += wmaenc.o wma.o
 OBJS-$(CONFIG_WMAV2_DECODER)           += wmadec.o wma.o
 OBJS-$(CONFIG_WMAV2_ENCODER)           += wmaenc.o wma.o
+OBJS-$(CONFIG_WMAPRO_DECODER)          += wmaprodec.o wma.o
 OBJS-$(CONFIG_WMV1_DECODER)            += h263dec.o h263.o mpeg12data.o mpegvideo.o error_resilience.o
 OBJS-$(CONFIG_WMV1_ENCODER)            += mpegvideo_enc.o motion_est.o ratecontrol.o h263.o mpeg12data.o mpegvideo.o error_resilience.o
 OBJS-$(CONFIG_WMV2_DECODER)            += wmv2dec.o wmv2.o msmpeg4.o msmpeg4data.o h263dec.o h263.o intrax8.o intrax8dsp.o mpeg12data.o mpegvideo.o error_resilience.o
Index: doc/general.texi
===================================================================
--- doc/general.texi	(revision 19340)
+++ doc/general.texi	(working copy)
@@ -632,6 +632,7 @@
 @item Westwood Audio (SND1)  @tab     @tab  X
 @item Windows Media Audio 1  @tab  X  @tab  X
 @item Windows Media Audio 2  @tab  X  @tab  X
+@item Windows Media Audio Pro @tab     @tab  X
 @end multitable
 
 @code{X} means that encoding (resp. decoding) is supported.
Index: MAINTAINERS
===================================================================
--- MAINTAINERS	(revision 19340)
+++ MAINTAINERS	(working copy)
@@ -200,6 +200,7 @@
   vp6                                   Aurelien Jacobs
   vqavideo.c                            Mike Melanson
   wavpack.c                             Kostya Shishkov
+  wmaprodec.c                           Sascha Sommer
   wmv2.c                                Michael Niedermayer
   wnv1.c                                Kostya Shishkov
   xan.c                                 Mike Melanson
