diff --git a/CMakeLists.txt b/CMakeLists.txt
index 6932b6fab998..61c95444fd18 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -210,7 +210,7 @@ set(CMAKE_RUNTIME_OUTPUT_DIRECTORY ${PROJECT_BINARY_DIR}/bin)
 # =======================================================================
 
 # Enforce the search mode of non-required packages for better and shorter failure messages
-find_package(Boost 1.73.0 REQUIRED context)
+find_package(Boost 1.81.0 REQUIRED context)
 find_package(enet 1.3 MODULE)
 find_package(fmt 9 REQUIRED)
 find_package(inih 52 MODULE COMPONENTS INIReader)
diff --git a/externals/vcpkg b/externals/vcpkg
index 9b22b40c6c61..a7b6122f6b65 160000
--- a/externals/vcpkg
+++ b/externals/vcpkg
@@ -1 +1 @@
-Subproject commit 9b22b40c6c61bf0da2d46346dd44a11e90972cc9
+Subproject commit a7b6122f6b6504d16d96117336a0562693579933
diff --git a/src/common/zstd_compression.cpp b/src/common/zstd_compression.cpp
index b71a41b78c36..cb6ec171b5fd 100644
--- a/src/common/zstd_compression.cpp
+++ b/src/common/zstd_compression.cpp
@@ -33,7 +33,7 @@ std::vector<u8> CompressDataZSTDDefault(const u8* source, std::size_t source_siz
 
 std::vector<u8> DecompressDataZSTD(std::span<const u8> compressed) {
     const std::size_t decompressed_size =
-        ZSTD_getDecompressedSize(compressed.data(), compressed.size());
+        ZSTD_getFrameContentSize(compressed.data(), compressed.size());
     std::vector<u8> decompressed(decompressed_size);
 
     const std::size_t uncompressed_result_size = ZSTD_decompress(
diff --git a/vcpkg.json b/vcpkg.json
index fbadca0e69d6..9aadd367cc40 100644
--- a/vcpkg.json
+++ b/vcpkg.json
@@ -1,7 +1,7 @@
 {
     "$schema": "https://raw.githubusercontent.com/microsoft/vcpkg-tool/main/docs/vcpkg.schema.json",
     "name": "yuzu",
-    "builtin-baseline": "9b22b40c6c61bf0da2d46346dd44a11e90972cc9",
+    "builtin-baseline": "a7b6122f6b6504d16d96117336a0562693579933",
     "version": "1.0",
     "dependencies": [
         "boost-algorithm",
