diff --git a/src/video_core/renderer_vulkan/vk_scheduler.cpp b/src/video_core/renderer_vulkan/vk_scheduler.cpp
index 55e699552df2..b264e6adabe9 100644
--- a/src/video_core/renderer_vulkan/vk_scheduler.cpp
+++ b/src/video_core/renderer_vulkan/vk_scheduler.cpp
@@ -328,7 +328,7 @@ void Scheduler::AcquireNewChunk() {
         chunk = std::make_unique<CommandChunk>();
     } else {
         // Otherwise, we can just take from the reserve.
-        chunk = std::make_unique<CommandChunk>();
+        chunk = std::move(chunk_reserve.back());
         chunk_reserve.pop_back();
     }
 }
