--- libvo/vo_vdpau.c.org	2010-11-16 01:42:05.109502215 -0500
+++ libvo/vo_vdpau.c	2010-11-16 01:44:32.108628079 -0500
@@ -148,7 +148,7 @@
 static int                                output_surface_width, output_surface_height;
 
 static VdpVideoMixer                      video_mixer;
-static int                                deint;
+static int                                deint, deint2=0;
 static int                                deint_type;
 static int                                deint_counter;
 static int                                pullup;
@@ -1053,6 +1053,21 @@
     else
         top_field_first = 1;
 
+    if (!deint2 && deint && !(mpi->fields & MP_IMGFIELD_INTERLACED))
+    {
+        deint2 = deint;
+        deint = 0;
+    }
+    else if (deint2 && !deint && (mpi->fields & MP_IMGFIELD_INTERLACED))
+    {
+        deint = deint2;
+        deint2 = 0;
+    }
+    else if (!deint && !deint2 && (mpi->fields & MP_IMGFIELD_INTERLACED))
+    {
+        deint = 3;
+    }
+
     video_to_output_surface();
     return VO_TRUE;
 }
