--- libvo/vo_vdpau.c.org	2010-02-15 00:16:57.000000000 -0500
+++ libvo/vo_vdpau.c	2010-02-15 00:18:37.000000000 -0500
@@ -184,6 +184,11 @@
 static int                                index_data_size;
 static uint32_t                           palette[PALETTE_SIZE];
 
+static int                                crop_top=0;
+static int                                crop_bottom=0;
+static int                                crop_left=0;
+static int                                crop_right=0;
+
 // EOSD
 // Pool of surfaces
 struct {
@@ -275,10 +280,12 @@
     out_rect_vid.x1 = dst_rect.right;
     out_rect_vid.y0 = dst_rect.top;
     out_rect_vid.y1 = dst_rect.bottom;
-    src_rect_vid.x0 = src_rect.left;
-    src_rect_vid.x1 = src_rect.right;
+    src_rect_vid.x0 = src_rect.left + crop_left;;
+    src_rect_vid.x1 = src_rect.right + crop_right;
     src_rect_vid.y0 = flip ? src_rect.bottom : src_rect.top;
+    src_rect_vid.y0 += crop_top;
     src_rect_vid.y1 = flip ? src_rect.top    : src_rect.bottom;
+    src_rect_vid.y1 += crop_bottom;
     border_x        = borders.left;
     border_y        = borders.top;
 #ifdef CONFIG_FREETYPE
@@ -1186,6 +1193,10 @@
     {"colorspace", OPT_ARG_INT, &colorspace, NULL},
     {"force-mixer", OPT_ARG_BOOL, &force_mixer, NULL},
     {"hqscaling", OPT_ARG_INT, &hqscaling, (opt_test_f)int_non_neg},
+    {"crop_top",    OPT_ARG_INT,  &crop_top,    NULL},
+    {"crop_bottom", OPT_ARG_INT,  &crop_bottom, NULL},
+    {"crop_left",   OPT_ARG_INT,  &crop_left,   NULL},
+    {"crop_right",  OPT_ARG_INT,  &crop_right,  NULL},
     {NULL}
 };
 
