--- src/util/videothumb.py.org	2010-12-12 12:15:04.843137002 -0500
+++ src/util/videothumb.py	2010-12-12 12:17:17.986053000 -0500
@@ -158,8 +158,8 @@
     os.chdir('/tmp')
 
     # call mplayer to get the image
-    _debug_('%r' % ((mplayer, '-nosound', '-vo', 'png', '-frames', '8', '-ss', position, '-zoom', filename),))
-    child = popen2.Popen3((mplayer, '-nosound', '-vo', 'png', '-frames', '8', '-ss', position, '-zoom', filename),
+    _debug_('%r' % ((mplayer, '-nosound', '-vc', 'ffh264,ffmpeg12,ffwmv3,ffvc1,', '-vo', 'png', '-frames', '8', '-ss', position, '-zoom', filename),))
+    child = popen2.Popen3((mplayer, '-nosound', '-vc', 'ffh264,ffmpeg12,ffwmv3,ffvc1,', '-vo', 'png', '-frames', '8', '-ss', position, '-zoom', filename),
         1, 100)
     while(1):
         data = child.fromchild.readline()
