--- xbmc/cores/dvdplayer/DVDCodecs/Audio/DVDAudioCodecFFmpeg.cpp.org	2010-04-24 15:30:44.744611299 -0400
+++ xbmc/cores/dvdplayer/DVDCodecs/Audio/DVDAudioCodecFFmpeg.cpp	2010-04-24 15:30:53.261735813 -0400
@@ -158,9 +158,6 @@
                                                  , pData
                                                  , iSize);
 
-#if (LIBAVCODEC_VERSION_INT > AV_VERSION_INT(52, 59, 0))
-  #error Make sure upstream version still needs this workaround (ffmpeg issue #1709)
-#endif
   /* upstream ac3dec is bugged, returns the packet size, not a negative value on error */
   if (m_pCodecContext->codec_id == CODEC_ID_AC3 && iBytesUsed > iSize)
   {
