--- tools/TexturePacker/Makefile.in.org	2014-05-25 18:15:22.638446666 -0700
+++ tools/TexturePacker/Makefile.in	2014-05-25 18:15:39.946447312 -0700
@@ -17,7 +17,8 @@
   -I. \
   -I@abs_top_srcdir@/lib \
   -I@abs_top_srcdir@/xbmc \
-  -I@abs_top_srcdir@/xbmc/linux
+  -I@abs_top_srcdir@/xbmc/linux \
+  $(shell pkg-config sdl SDL_image --cflags)
 
 LDFLAGS_FOR_BUILD += -lSDL_image -lSDL -llzo2
 LDFLAGS_FOR_BUILD += -L@abs_top_srcdir@/lib/libsquish -lsquish-native
