--- libavformat/utils.c.org	2012-02-03 23:18:35.000000000 -0500
+++ libavformat/utils.c	2012-02-03 23:19:09.000000000 -0500
@@ -3302,7 +3302,7 @@
         av_log(s, AV_LOG_ERROR,
                "Application provided invalid, non monotonically increasing dts to muxer in stream %d: %"PRId64" >= %"PRId64"\n",
                st->index, st->cur_dts, pkt->dts);
-        return AVERROR(EINVAL);
+//        return AVERROR(EINVAL);
     }
     if(pkt->dts != AV_NOPTS_VALUE && pkt->pts != AV_NOPTS_VALUE && pkt->pts < pkt->dts){
         av_log(s, AV_LOG_ERROR, "pts < dts in stream %d\n", st->index);
--- libavcodec/mpegvideo_enc.c.org	2012-02-03 23:18:28.000000000 -0500
+++ libavcodec/mpegvideo_enc.c	2012-02-03 23:18:53.000000000 -0500
@@ -968,7 +968,7 @@
                     av_log(s->avctx, AV_LOG_ERROR,
                            "Error, Invalid timestamp=%"PRId64", "
                            "last=%"PRId64"\n", pts, s->user_specified_pts);
-                    return -1;
+//                    return -1;
                 }
             }
             s->user_specified_pts = pts;
