--- libavformat/utils.c.org	2011-05-18 05:06:47.000000000 -0400
+++ libavformat/utils.c	2011-05-18 05:08:27.000000000 -0400
@@ -3000,7 +3000,7 @@
         av_log(s, AV_LOG_ERROR,
                "Application provided invalid, non monotonically increasing dts to muxer in stream %d: %"PRId64" >= %"PRId64"\n",
                st->index, st->cur_dts, pkt->dts);
-        return AVERROR(EINVAL);
+        //return AVERROR(EINVAL);
     }
     if(pkt->dts != AV_NOPTS_VALUE && pkt->pts != AV_NOPTS_VALUE && pkt->pts < pkt->dts){
         av_log(s, AV_LOG_ERROR, "pts < dts in stream %d\n", st->index);
--- libavcodec/mpegvideo_enc.c.org	2011-05-18 05:07:43.000000000 -0400
+++ libavcodec/mpegvideo_enc.c	2011-05-18 05:08:09.000000000 -0400
@@ -829,7 +829,7 @@
 
                 if(time <= last){
                     av_log(s->avctx, AV_LOG_ERROR, "Error, Invalid timestamp=%"PRId64", last=%"PRId64"\n", pts, s->user_specified_pts);
-                    return -1;
+                    //return -1;
                 }
             }
             s->user_specified_pts= pts;
