diff --git a/Makefile b/Makefile
index 6ed3be5..9728c4f 100644
--- a/Makefile
+++ b/Makefile
@@ -42,8 +42,8 @@ distclean:
 install:
 	$(INSTALL_PROGRAM) "./target/release/$(BIN_SD)" "$(DESTDIR)$(bindir)/$(BIN_SD)"
 	$(INSTALL_PROGRAM) "./target/release/$(BIN_SC)" "$(DESTDIR)$(bindir)/$(BIN_SC)"
-	$(INSTALL_DATA) "./data/$(SERVICE)" "$(DESTDIR)$(libdir)/systemd/system/$(SERVICE)"
-	$(INSTALL_DATA) "./data/$(PRESET)" "$(DESTDIR)$(libdir)/systemd/system-preset/$(PRESET)"
+	$(INSTALL_DATA) "./data/$(SERVICE)" "$(DESTDIR)/lib/systemd/system/$(SERVICE)"
+	$(INSTALL_DATA) "./data/$(PRESET)" "$(DESTDIR)/lib//systemd/system-preset/$(PRESET)"
 	$(INSTALL_DATA) "./data/$(DBUSCFG)" "$(DESTDIR)$(datarootdir)/dbus-1/system.d/$(DBUSCFG)"
 	$(INSTALL_DATA) "./data/$(X11CFG)" "$(DESTDIR)$(datarootdir)/X11/xorg.conf.d/$(X11CFG)"
 	$(INSTALL_DATA) "./data/$(PMRULES)" "$(DESTDIR)$(libdir)/udev/rules.d/$(PMRULES)"
