--- systems/cardclient/newcamd.c.org	2009-08-18 23:29:17.000000000 -0400
+++ systems/cardclient/newcamd.c	2009-08-18 23:31:44.000000000 -0400
@@ -506,13 +506,16 @@
 bool cCardClientNewCamd::ProcessEMM(int caSys, const unsigned char *data)
 {
   if(emmProcessing && emmAllowed) {
+    LDUMP(L_CC_NEWCAMD,data,16,"EMM in");
     cMutexLock lock(this);
     cAssembleData ad(data);
     if(MatchAndAssemble(&ad,0,0)) {
+      PRINTF(L_CC_NEWCAMD,"matched");
       while((data=ad.Assembled())) {
         int len=SCT_LEN(data);
         int id=msEMM.Get(data,len,0);
         if(id>0) {
+          LDUMP(L_CC_NEWCAMD,data,24,"EMM out");
           if(SendMessage(&so,data,len,true,0)) {
             unsigned char buffer[CWS_NETMSGSIZE];
             len=ReceiveMessage(&so,buffer,true);
