--- cam.c.org	2009-05-02 12:50:28.000000000 -0400
+++ cam.c	2009-05-02 12:50:44.000000000 -0400
@@ -747,7 +747,7 @@
   bool UpdateEcm(void);
   void EcmOk(void);
   void EcmFail(void);
-  void ParseCAInfo(int sys);
+  void ParseCAInfo(int sys, int source=0);
   void AddEcmPri(cEcmInfo *n);
 protected:
   virtual void Process(cPidFilter *filter, unsigned char *data, int len);
@@ -1156,7 +1156,7 @@
     bool log=dolog;
     dolog=(sys && sys->NeedsData() && ecm->Data()==0);
     if(dolog) PRINTF(L_CORE_ECM,"%s: try to update ecm extra data",id);
-    ParseCAInfo(ecm->caId);
+    ParseCAInfo(ecm->caId, ecm->source);
     ecmUpdTime.Set(ECM_UPD_TIME);
     dolog=log;
     if(!ecm->Data()) return false;
@@ -1218,7 +1218,7 @@
   ecm->Fail(true);
 }
 
-void cEcmHandler::ParseCAInfo(int SysId)
+void cEcmHandler::ParseCAInfo(int SysId, int Source)
 {
   unsigned char buff[2048];
   caid_t casys[MAXCAIDS+1];
@@ -1246,7 +1246,7 @@
         while((sys=cSystems::FindBySysId(sysId,!cam->IsSoftCSA(),sysPri))) {
           sysPri=sys->Pri();
           cSimpleList<cEcmInfo> ecms;
-          sys->ParseCADescriptor(&ecms,sysId,&buff[index+2],buff[index+1]);
+          sys->ParseCADescriptor(&ecms,sysId,Source,&buff[index+2],buff[index+1]);
           delete sys;
           if(ecms.Count()) {
             cEcmInfo *n;
--- system.c.org	2009-05-02 12:50:17.000000000 -0400
+++ system.c	2009-05-02 12:51:28.000000000 -0400
@@ -126,7 +126,7 @@
     }
 }
 
-void cSystem::ParseCADescriptor(cSimpleList<cEcmInfo> *ecms, unsigned short sysId, const unsigned char *data, int len)
+void cSystem::ParseCADescriptor(cSimpleList<cEcmInfo> *ecms, unsigned short sysId, unsigned short source, const unsigned char *data, int len)
 {
   const int pid=WORD(data,2,0x1FFF);
   switch(sysId>>8) {
@@ -145,7 +145,7 @@
     default:   // default style
       {
       cEcmInfo *n=new cEcmInfo(name,pid,sysId,0);
-      if(sysId==0x1234) { // BEV
+      if((sysId==0x1234) || (source==0x8334) || (source==0x838e)) { // BEV
         n->ecm_table=0x8e;
         n->emmCaId=0x1801;
         }
--- system.h.org	2009-05-02 12:50:23.000000000 -0400
+++ system.h	2009-05-02 12:50:44.000000000 -0400
@@ -117,7 +117,7 @@
   virtual void CheckECMResult(const cEcmInfo *ecm, const unsigned char *data, bool result);
   virtual bool ProcessECM(const cEcmInfo *ecm, unsigned char *buffer)=0;
   virtual void ProcessEMM(int pid, int caid, unsigned char *buffer) {};
-  virtual void ParseCADescriptor(cSimpleList<cEcmInfo> *ecms, unsigned short sysId, const unsigned char *data, int len);
+  virtual void ParseCADescriptor(cSimpleList<cEcmInfo> *ecms, unsigned short sysId, unsigned short source, const unsigned char *data, int len);
   virtual void ParseCAT(cPids *pids, const unsigned char *buffer);
   unsigned char *CW(void) { return cw; }
   void DoLog(bool Log) { doLog=Log; }
