--- sc.c.org	2009-05-20 15:09:30.000000000 -0400
+++ sc.c	2009-05-20 15:13:41.000000000 -0400
@@ -1193,6 +1193,7 @@
   char pat[32];
   snprintf(pat,sizeof(pat),"%s*-%d%s%s",LIBSC_PREFIX,SCAPIVERS,SO_INDICATOR,APIVERSION);
   bool res=true;
+/*
   cReadDir dir(path);
   struct dirent *e;
   while((e=dir.Next())) {
@@ -1205,6 +1206,23 @@
       }
     }
   free(path);
+*/
+  DIR *dir = opendir (path);
+  struct dirent *e;
+
+  while ((e = readdir (dir)) != NULL) {
+    if(!fnmatch(pat,e->d_name,FNM_PATHNAME|FNM_NOESCAPE)) {
+      cScDll *dll=new cScDll(AddDirectory(path,e->d_name));
+      if(dll) {
+        if(!dll->Load()) res=false;
+        Ins(dll);
+        }
+      }
+  }
+
+  closedir (dir);
+  free(path);
+
   return res;
 }
 
