--- libmpdemux/demuxer.c.org	2012-01-03 23:42:42.215787543 -0500
+++ libmpdemux/demuxer.c	2012-01-03 23:41:22.639943978 -0500
@@ -854,7 +854,7 @@
 struct demux_packet *ds_get_packet2(struct demux_stream *ds)
 {
     // This shouldn't get used together with partial reads
-    assert(ds->buffer_pos >= ds->buffer_size);
+    //assert(ds->buffer_pos >= ds->buffer_size);
     ds_fill_buffer(ds);
     ds->buffer_pos = ds->buffer_size;
     return ds->current;
