--- mplayer.c.org	2011-04-15 00:46:24.000000000 -0400
+++ mplayer.c	2011-04-15 01:34:57.000000000 -0400
@@ -2889,6 +2889,23 @@
 
     while ( (cmd = mp_input_get_cmd(mpctx->input, 20, 1)) == NULL
             || cmd->id == MP_CMD_SET_MOUSE_POS || cmd->pausing == 4) {
+    if (live_pause_enable)
+    {
+        off_t sz = 0;
+		struct MPOpts *opts = &mpctx->opts;
+        sh_video_t *sh_video = mpctx->demuxer->video->sh;
+		double new_pts = 0;
+        (*mpctx->stream->control)(mpctx->stream, STREAM_CTRL_GET_SIZE, &sz);
+
+		new_pts = (double) (sz - mpctx->demuxer->filepos) / sh_video->pts;
+
+        set_osd_msg(OSD_MSG_TEXT, 1, opts->osd_duration, "Buffering.. %0.2f", new_pts);
+
+        printf("HI0: %ld -- %ld (%0.2f)\n", sz, mpctx->demuxer->filepos, new_pts);
+		if(new_pts > live_pause_enable)
+               	unpause_player(mpctx);
+    }
+
 	if (cmd) {
             cmd = mp_input_get_cmd(mpctx->input, 0, 0);
             run_command(mpctx, cmd);
@@ -3474,6 +3491,23 @@
             }
         }
 
+        if (live_pause_enable && mpctx->stream && mpctx->demuxer)
+        {
+
+            off_t sz = 0;
+            sh_video_t *sh_video = mpctx->demuxer->video->sh;
+			if (sh_video->pts > 0 && first_pts_hack > 0 && sh_video->pts > first_pts_hack)
+			{
+	            double new_pts = 0;
+        	    (*mpctx->stream->control)(mpctx->stream, STREAM_CTRL_GET_SIZE, &sz);
+				new_pts = (double) (sz - mpctx->demuxer->filepos) / sh_video->pts;
+        	    printf("HI: %ld %ld (%0.2f)\n", sz, mpctx->demuxer->filepos, new_pts);
+				if(new_pts < live_pause_enable)
+                	pause_player(mpctx);
+			}
+        }
+
+
         bool frame_time_remaining = sleep_until_near_frame(mpctx,
                                                            &mpctx->time_frame,
                                                            &aq_sleep_time);
