--- xbmc-20.1-Nexus/xbmc/cores/RetroPlayer/cheevos/Cheevos.h.org	2023-05-05 03:16:07.893921045 -0400
+++ xbmc-20.1-Nexus/xbmc/cores/RetroPlayer/cheevos/Cheevos.h	2023-05-05 03:16:31.554052559 -0400
@@ -12,6 +12,7 @@
 
 #include <map>
 #include <string>
+#include <cinttypes>
 
 namespace KODI
 {
--- xbmc-20.1-Nexus/xbmc/cores/VideoPlayer/VideoRenderers/ColorManager.h.org	2023-05-05 03:24:52.508062730 -0400
+++ xbmc-20.1-Nexus/xbmc/cores/VideoPlayer/VideoRenderers/ColorManager.h	2023-05-05 03:25:07.278180406 -0400
@@ -13,6 +13,7 @@
 #endif
 
 #include <string>
+#include <cinttypes>
 
 extern "C"
 {
--- xbmc-20.1-Nexus/xbmc/cores/VideoPlayer/VideoRenderers/HwDecRender/VaapiEGL.h.org	2023-05-05 03:25:49.958494057 -0400
+++ xbmc-20.1-Nexus/xbmc/cores/VideoPlayer/VideoRenderers/HwDecRender/VaapiEGL.h	2023-05-05 03:26:14.228636409 -0400
@@ -9,6 +9,7 @@
 #pragma once
 
 #include <array>
+#include <cinttypes>
 
 #if defined(HAS_GL)
 // always define GL_GLEXT_PROTOTYPES before include gl headers
--- xbmc-20.1-Nexus/xbmc/pictures/Picture.h.org	2023-05-05 03:19:33.698549847 -0400
+++ xbmc-20.1-Nexus/xbmc/pictures/Picture.h	2023-05-05 03:19:50.552051679 -0400
@@ -13,6 +13,7 @@
 
 #include <string>
 #include <vector>
+#include <cinttypes>
 
 extern "C"
 {
--- xbmc-20.1-Nexus/xbmc/platform/Filesystem.h.org	2023-05-05 03:26:37.158770887 -0400
+++ xbmc-20.1-Nexus/xbmc/platform/Filesystem.h	2023-05-05 03:26:53.102197738 -0400
@@ -9,6 +9,7 @@
 #pragma once
 
 #include <string>
+#include <cinttypes>
 #include <system_error>
 namespace KODI
 {
--- xbmc-20.1-Nexus/xbmc/utils/CSSUtils.h.org	2023-05-05 03:23:01.087140328 -0400
+++ xbmc-20.1-Nexus/xbmc/utils/CSSUtils.h	2023-05-05 03:23:17.910617189 -0400
@@ -9,6 +9,7 @@
 #pragma once
 
 #include <string>
+#include <cinttypes>
 
 namespace UTILS
 {
--- xbmc-20.1-Nexus/xbmc/utils/CharArrayParser.h.org	2023-05-05 03:21:48.893171548 -0400
+++ xbmc-20.1-Nexus/xbmc/utils/CharArrayParser.h	2023-05-05 03:22:02.206624541 -0400
@@ -9,6 +9,7 @@
 #pragma once
 
 #include <string>
+#include <cinttypes>
 
 /*!
  *\brief Wraps a char array, providing a set of methods for parsing data from it.
--- xbmc-20.1-Nexus/xbmc/utils/StreamUtils.h.org	2023-05-05 03:24:19.951133169 -0400
+++ xbmc-20.1-Nexus/xbmc/utils/StreamUtils.h	2023-05-05 03:24:32.277899981 -0400
@@ -9,6 +9,7 @@
 #pragma once
 
 #include <string>
+#include <cinttypes>
 
 static constexpr int MP4_BOX_HEADER_SIZE = 8;
 
--- xbmc-20.1-Nexus/xbmc/utils/DRMHelpers.h.org	2023-05-05 03:50:52.777247140 -0400
+++ xbmc-20.1-Nexus/xbmc/utils/DRMHelpers.h	2023-05-05 03:51:10.247348260 -0400
@@ -9,6 +9,7 @@
 #pragma once
 
 #include <string>
+#include <cinttypes>
 
 namespace DRMHELPERS
 {
