--- /usr/portage/media-tv/kodi/kodi-16.0.ebuild	2016-03-03 13:53:05.000000000 -0500
+++ kodi-16.0.ebuild	2016-02-20 16:20:55.230882292 -0500
@@ -1,27 +1,37 @@
-# Copyright 1999-2016 Gentoo Foundation
+
+# Copyright 1999-2015 Gentoo Foundation
 # Distributed under the terms of the GNU General Public License v2
 # $Id$
 
 EAPI="5"
 
 # Does not work with py3 here
+# It might work with py:2.5 but I didn't test that
 PYTHON_COMPAT=( python2_7 )
 PYTHON_REQ_USE="sqlite"
 
 inherit eutils linux-info python-single-r1 multiprocessing autotools toolchain-funcs
-
 CODENAME="Jarvis"
 case ${PV} in
 9999)
 	EGIT_REPO_URI="git://github.com/xbmc/xbmc.git"
 	inherit git-r3
 	;;
+*_alpha*|*_beta*|*_rc*)
+	MY_PV=${PV/_alpha/a}
+	MY_PV=${MY_PV/_beta/b}
+	MY_PV=${MY_PV/_rc/rc}
+	MY_P="${PN}${MY_PV}"
+	#MY_P="${PN}${MY_PV}"
+	SRC_URI="https://github.com/xbmc/xbmc/archive/${MY_PV}-${CODENAME}.tar.gz -> ${P}.tar.gz"
+	KEYWORDS="~amd64 ~x86"
+	S=${WORKDIR}/xbmc-${MY_PV}-${CODENAME}
+	;;
 *|*_p*)
 	MY_PV=${PV/_p/_r}
 	MY_P="${PN}-${MY_PV}"
 	SRC_URI="http://mirrors.kodi.tv/releases/source/${MY_PV}-${CODENAME}.tar.gz -> ${P}.tar.gz
-		https://github.com/xbmc/xbmc/archive/${PV}-${CODENAME}.tar.gz -> ${P}.tar.gz
-		!java? ( http://mirrors.kodi.tv/releases/source/${MY_P}-generated-addons.tar.xz )"
+		https://github.com/xbmc/xbmc/archive/${PV}-${CODENAME}.tar.gz -> ${P}.tar.gz"
 	KEYWORDS="~amd64 ~x86"
 
 	S=${WORKDIR}/xbmc-${PV}-${CODENAME}
@@ -29,16 +39,12 @@
 esac
 
 DESCRIPTION="Kodi is a free and open source media-player and entertainment hub"
-HOMEPAGE="https://kodi.tv/ http://kodi.wiki/"
+HOMEPAGE="http://kodi.tv/ http://kodi.wiki/"
 
 LICENSE="GPL-2"
 SLOT="0"
-IUSE="airplay alsa avahi bluetooth bluray caps cec css dbus debug gles java joystick midi mysql nfs +opengl profile pulseaudio rtmp +samba sftp test +texturepacker udisks upnp upower +usb vaapi vdpau webserver +X"
-# gles/vaapi: http://trac.kodi.tv/ticket/10552 #464306
+IUSE="airplay alsa avahi bluetooth bluray caps cec css dbus debug gles java joystick midi mysql nfs +opengl profile pulseaudio rtmp +samba sftp +spectrum test +texturepacker udisks upnp upower +usb vaapi vdpau webserver +X -external_ffmpeg"
 REQUIRED_USE="
-	|| ( gles opengl )
-	gles? ( !vaapi )
-	vaapi? ( !gles )
 	udisks? ( dbus )
 	upower? ( dbus )
 "
@@ -50,6 +56,7 @@
 	app-i18n/enca
 	airplay? ( app-pda/libplist )
 	dev-libs/boost
+	dev-libs/crossguid
 	dev-libs/expat
 	dev-libs/fribidi
 	dev-libs/libcdio[-minimal]
@@ -59,7 +66,7 @@
 	dev-libs/libxslt
 	>=dev-libs/lzo-2.04
 	dev-libs/tinyxml[stl]
-	>=dev-libs/yajl-2
+	dev-libs/yajl
 	dev-python/simplejson[${PYTHON_USEDEP}]
 	media-fonts/corefonts
 	media-fonts/roboto
@@ -70,7 +77,7 @@
 	media-libs/jasper
 	media-libs/jbigkit
 	>=media-libs/libass-0.9.7
-	bluray? ( >=media-libs/libbluray-0.7.0 )
+	bluray? ( media-libs/libbluray )
 	css? ( media-libs/libdvdcss )
 	media-libs/libmad
 	media-libs/libmodplug
@@ -82,10 +89,10 @@
 	>=media-libs/taglib-1.8
 	media-libs/libvorbis
 	media-libs/tiff:0=
-	media-sound/dcadec
 	pulseaudio? ( media-sound/pulseaudio )
 	media-sound/wavpack
-	>=media-video/ffmpeg-2.6:=[encode]
+	media-sound/dcadec
+	external_ffmpeg? ( >=media-video/ffmpeg-2.6:=[encode] )
 	rtmp? ( media-video/rtmpdump )
 	avahi? ( net-dns/avahi )
 	nfs? ( net-fs/libnfs:= )
@@ -110,7 +117,7 @@
 	)
 	vaapi? ( x11-libs/libva[opengl] )
 	vdpau? (
-		|| ( >=x11-libs/libvdpau-1.1 >=x11-drivers/nvidia-drivers-180.51 )
+		|| ( x11-libs/libvdpau >=x11-drivers/nvidia-drivers-180.51 )
 		media-video/ffmpeg[vdpau]
 	)
 	X? (
@@ -127,15 +134,13 @@
 DEPEND="${COMMON_DEPEND}
 	app-arch/xz-utils
 	dev-lang/swig
-	dev-libs/crossguid
 	dev-util/gperf
 	texturepacker? ( media-libs/giflib )
 	X? ( x11-proto/xineramaproto )
 	dev-util/cmake
 	x86? ( dev-lang/nasm )
 	java? ( virtual/jre )
-	test? ( dev-cpp/gtest )
-	virtual/pkgconfig"
+	test? ( dev-cpp/gtest )"
 # Force java for latest git version to avoid having to hand maintain the
 # generated addons package.  #488118
 [[ ${PV} == "9999" ]] && DEPEND+=" virtual/jre"
@@ -157,8 +162,7 @@
 
 src_prepare() {
 	epatch "${FILESDIR}"/${PN}-9999-no-arm-flags.patch #400617
-	epatch "${FILESDIR}"/${PN}-9999-texturepacker.patch
-	epatch "${FILESDIR}"/${PN}-16-ffmpeg3.patch
+	epatch "${FILESDIR}"/${PN}-15.1-texturepacker.patch
 	epatch_user #293109
 
 	# some dirs ship generated autotools, some dont
@@ -194,11 +198,27 @@
 		-e '/dbus_connection_send_with_reply_and_block/s:-1:3000:' \
 		xbmc/linux/*.cpp || die
 
+	case ${PV} in
+		*|*_p*)
+		S=${WORKDIR}/xbmc-${MY_PV}-${CODENAME}
+		cd $S/tools/depends/native/JsonSchemaBuilder/src
+		./autogen.sh
+		./configure
+		make
+		cd ..
+		make
+		;;
+	esac
+
 	# Tweak autotool timestamps to avoid regeneration
 	find . -type f -exec touch -r configure {} +
 }
 
 src_configure() {
+	FFMPEG_TYPE=force
+	if use external_ffmpeg; then
+		FFMPEG_TYPE=shared
+	fi
 	# Disable documentation generation
 	export ac_cv_path_LATEX=no
 	# Avoid help2man
@@ -212,7 +232,7 @@
 		--docdir=/usr/share/doc/${PF} \
 		--disable-ccache \
 		--disable-optimizations \
-		--with-ffmpeg=shared \
+		--with-ffmpeg=$FFMPEG_TYPE \
 		$(use_enable alsa) \
 		$(use_enable airplay) \
 		$(use_enable avahi) \
@@ -249,18 +269,30 @@
 
 src_install() {
 	default
-	rm "${ED}"/usr/share/doc/*/{LICENSE.GPL,copying.txt}* || die
+	rm "${ED}"/usr/share/doc/*/{LICENSE.GPL,copying.txt}*
 
 	domenu tools/Linux/kodi.desktop
 	newicon media/icon48x48.png kodi.png
 
+	# Remove optional addons (platform specific).
+	local disabled_addons=(
+		repository.pvr-{android,ios,osx{32,64},win32}.xbmc.org
+		visualization.dxspectrum
+		visualization.vortex
+	)
+	rm -rf "${disabled_addons[@]/#/${ED}/usr/share/kodi/addons/}"
+
 	# Remove fonconfig settings that are used only on MacOSX.
 	# Can't be patched upstream because they just find all files and install
 	# them into same structure like they have in git.
 	rm -rf "${ED}"/usr/share/kodi/system/players/dvdplayer/etc
 
-	# Replace bundled fonts with system ones.
-	rm "${ED}"/usr/share/kodi/addons/skin.confluence/fonts/Roboto-* || die
+	# Replace bundled fonts with system ones
+	# teletext.ttf: unknown
+	# bold-caps.ttf: unknown
+	# roboto: roboto-bold, roboto-regular
+	# arial.ttf: font mashed from droid/roboto, not removed wrt bug#460514
+	rm -rf "${ED}"/usr/share/kodi/addons/skin.confluence/fonts/Roboto-*
 	dosym /usr/share/fonts/roboto/Roboto-Regular.ttf \
 		/usr/share/kodi/addons/skin.confluence/fonts/Roboto-Regular.ttf
 	dosym /usr/share/fonts/roboto/Roboto-Bold.ttf \
