--- src/www/htdocs/manualrecord.rpy.org	2009-10-21 21:17:15.000000000 -0400
+++ src/www/htdocs/manualrecord.rpy	2009-10-21 21:19:55.000000000 -0400
@@ -29,18 +29,11 @@
 # -----------------------------------------------------------------------
 
 import sys, time
+from time import strptime
 
 import tv.epg_xmltv, tv.epg_types
 from tv.record_client import RecordClient
 
-# Use the alternate strptime module which seems to handle time zones
-#
-# XXX Remove when we are ready to require Python 2.3
-if float(sys.version[0:3]) < 2.3:
-    import tv.strptime as strptime
-else:
-    import _strptime as strptime
-
 
 from www.web_types import HTMLResource, FreevoResource
 
@@ -103,10 +96,10 @@
                 if int(startmonth) < currentmonth:
                     startyear = str(int(startyear) + 1)
                 # create utc second start time
-                starttime = time.mktime(strptime.strptime(str(startmonth)+' '+str(startday)+' '+str(startyear)+\
+                starttime = time.mktime(strptime(str(startmonth)+' '+str(startday)+' '+str(startyear)+\
                     ' '+str(starthour)+':'+str(startminute)+':00','%m %d %Y %H:%M:%S'))
                 # create utc stop time
-                stoptime = time.mktime(strptime.strptime(str(stopmonth)+' '+str(stopday)+' '+str(stopyear)+\
+                stoptime = time.mktime(strptime(str(stopmonth)+' '+str(stopday)+' '+str(stopyear)+\
                     ' '+str(stophour)+':'+str(stopminute)+':00','%m %d %Y %H:%M:%S'))
                 # so we don't record for more then maxdays (maxdays day * 24hr/day * 60 min/hr * 60 sec/min)
                 if abs(stoptime - starttime) < (MAXDAYS * 86400):
