--- src/helpers/recordserver.py.org	2010-10-27 00:05:36.046162999 -0400
+++ src/helpers/recordserver.py	2010-10-27 00:13:49.251162999 -0400
@@ -884,7 +884,8 @@
                 recording = False
 
             if recording:
-                currently_recording = prog
+                currently_recording = None
+                #currently_recording = prog
 
         now = time.time()
         for prog in progs.values():
@@ -1330,7 +1331,7 @@
             elif event == RECORD_START:
                 prog = event.arg
                 _debug_('RECORD_START %s' % (prog), DINFO)
-                open(self.tv_lock_file, 'w').close()
+                #open(self.tv_lock_file, 'w').close()
                 self.create_fxd(prog)
                 if config.VCR_PRE_REC:
                     util.popen3.Popen3(config.VCR_PRE_REC)
@@ -1443,7 +1444,7 @@
 
 
     _debug_('kaa.AtTimer starting')
-    kaa.AtTimer(recordserver.handleAtTimer).start(sec=45)
+    kaa.AtTimer(recordserver.handleAtTimer).start(sec=0)
     _debug_('kaa.main starting')
     kaa.main.run()
     _debug_('kaa.main finished')
