--- src/helpers/recordserver.py.org	2010-11-15 15:57:40.778999994 -0500
+++ src/helpers/recordserver.py	2010-11-15 15:58:24.418999993 -0500
@@ -896,7 +896,8 @@
         for prog in progs.values():
             recording = hasattr(prog, 'isRecording') and prog.isRecording
             if recording:
-                currently_recording = prog
+                currently_recording = None
+                #currently_recording = prog
 
         if currently_recording is not None:
             _debug_('currently_recording=%s' % (currently_recording), DINFO)
@@ -1016,7 +1017,7 @@
             if event == RECORD_START:
                 prog = event.arg
                 _debug_('RECORD_START %s' % (prog), DINFO)
-                open(self.tv_lockfile, 'w').close()
+                #open(self.tv_lockfile, 'w').close()
                 self.create_fxd(prog)
                 if config.VCR_PRE_REC:
                     util.popen3.Popen3(config.VCR_PRE_REC)
