--- freevo.org	2010-11-16 21:20:38.000000000 -0500
+++ freevo	2010-11-16 21:21:50.000000000 -0500
@@ -33,7 +33,7 @@
 import os
 import sys
 import time
-import popen2
+from subprocess import Popen, PIPE
 
 from stat import *
 from signal import *
@@ -123,26 +123,22 @@
                 # try to import freevo with this python and get
                 # the path
                 cmd = '%s -c "import freevo; print freevo.__path__[0]"' % python
-                child = popen2.Popen4(cmd)
+                child = Popen(cmd, shell=True, stdin=PIPE, stdout=PIPE)
                 while 1:
-                    data = child.fromchild.readline()
+                    data = child.stdout.readline()
                     if not data:
                         break
                     if os.path.isdir(data[:-1]):
                         # ok, found it, close child and return
+                        child.stdout.close()
+                        child.stdin.close()
                         child.wait()
-                        child.fromchild.close()
-                        if child.childerr:
-                            child.childerr.close()
-                        child.tochild.close()
-                        _debug_('python=%r data=%r' % (python, data[:-1])) 
+                        _debug_('python=%r data=%r' % (python, data[:-1]))
                         return python, data[:-1]
 
                 child.wait()
-                child.fromchild.close()
-                if child.childerr:
-                    child.childerr.close()
-                child.tochild.close()
+                child.stdout.close()
+                child.stdin.close()
 
     # nothing found? That's bad!
     if check_freevo:
--- src/helpers/recordserver.py.org	2010-11-16 21:13:23.000000000 -0500
+++ src/helpers/recordserver.py	2010-11-16 21:13:30.000000000 -0500
@@ -26,7 +26,7 @@
 # -----------------------------------------------------------------------
 
 
-import sys, string, random, time, os, re, pwd, stat, threading, md5, datetime, copy
+import sys, string, random, time, os, re, pwd, stat, threading, hashlib, datetime, copy
 try:
     import cPickle as pickle
 except ImportError:
