--- stream/stream_ffmpeg.c.org	2012-11-04 23:54:31.202380663 -0500
+++ stream/stream_ffmpeg.c	2012-11-04 23:55:31.796408775 -0500
@@ -117,6 +117,13 @@
     for (int i = 0; i < sizeof(prefix) / sizeof(prefix[0]); i++)
         if (!strncmp(filename, prefix[i], strlen(prefix[i])))
             filename += strlen(prefix[i]);
+    if (!strncmp(filename, "applehttp:", 10)) {
+        char *newname = strdup(&stream->url[5]);
+        if (newname) {
+            free(stream->url);
+            filename = stream->url = newname;
+        }
+    }
     if (!strncmp(filename, "rtsp:", 5)) {
         /* This is handled as a special demuxer, without a separate
          * stream layer. demux_lavf will do all the real work.
@@ -164,7 +171,7 @@
   "",
   "",
   open_f,
-  { "lavf", "ffmpeg", "rtmp", "rtsp", NULL },
+  { "lavf", "ffmpeg", "rtmp", "applehttp", "rtsp", NULL },
   NULL,
   1 // Urls are an option string
 };
