--- recorder.c.org	2011-02-24 19:08:53.000000000 -0500
+++ recorder.c	2011-02-24 19:09:29.000000000 -0500
@@ -10,7 +10,7 @@
 #include "recorder.h"
 #include "shutdown.h"
 
-#define RECORDERBUFSIZE  MEGABYTE(5)
+#define RECORDERBUFSIZE  MEGABYTE(200)
 
 // The maximum time we wait before assuming that a recorded video data stream
 // is broken:
--- dvbplayer.c.org	2011-02-24 19:09:01.000000000 -0500
+++ dvbplayer.c	2011-02-24 19:14:45.000000000 -0500
@@ -191,9 +191,9 @@
 
 // --- cDvbPlayer ------------------------------------------------------------
 
-#define PLAYERBUFSIZE  MEGABYTE(1)
+#define PLAYERBUFSIZE  MEGABYTE(70)
 
-#define RESUMEBACKUP 10 // number of seconds to back up when resuming an interrupted replay session
+#define RESUMEBACKUP 20 // number of seconds to back up when resuming an interrupted replay session
 #define MAXSTUCKATEOF 3 // max. number of seconds to wait in case the device doesn't play the last frame
 
 class cDvbPlayer : public cPlayer, cThread {
--- dvbdevice.c.org	2011-02-24 20:26:24.000000000 -0500
+++ dvbdevice.c	2011-02-24 20:26:47.000000000 -0500
@@ -707,7 +707,7 @@
   CloseDvr();
   fd_dvr = DvbOpen(DEV_DVB_DVR, adapter, frontend, O_RDONLY | O_NONBLOCK, true);
   if (fd_dvr >= 0)
-     tsBuffer = new cTSBuffer(fd_dvr, MEGABYTE(2), CardIndex() + 1);
+     tsBuffer = new cTSBuffer(fd_dvr, MEGABYTE(10), CardIndex() + 1);
   return fd_dvr >= 0;
 }
 
